lucene/core/src/java/org/apache/lucene/document/IndexSortDocValuesRangeQuery.java

```diff
@@ -0,0 +1,279 @@
+                                                      long upperValue,
+                                                      SortField sortField,
+                                                      LeafReaderContext context,
+                                                      DocIdSetIterator delegate) throws IOException {
+    int maxDoc = context.reader().maxDoc();
+    if (maxDoc <= 0) {
+      return DocIdSetIterator.empty();
+    }
+
+    if (sortField.getReverse()) {
+      long temp = lowerValue;
+      lowerValue = upperValue;
+      upperValue = temp;
+    }
+
+    // Perform a binary search to find the first document with value >= lowerValue.
+    ValueComparator comparator = loadComparator(sortField, lowerValue, context);
+    int low = 0;
+    int high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      if (comparator.compare(mid) <= 0) {
+        high = mid - 1;
+        comparator = loadComparator(sortField, lowerValue, context);
+      } else {
+        low = mid + 1;
+      }
+    }
+    int firstDocIdInclusive = high + 1;
+
+    // Perform a binary search to find the first document with value > upperValue.
+    // Since we know that upperValue >= lowerValue, we can initialize the lower bound
+    // of the binary search to the result of the previous search.
+    comparator = loadComparator(sortField, upperValue, context);
+    low = firstDocIdInclusive;
+    high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      if (comparator.compare(mid) < 0) {
+        high = mid - 1;
+        comparator = loadComparator(sortField, upperValue, context);
+      } else {
+        low = mid + 1;
+      }
+    }
+
+    int lastDocIdExclusive = high + 1;
+    return new BoundedDocSetIdIterator(firstDocIdInclusive, lastDocIdExclusive, delegate);
+  }
+
+  /**
+   * Compares the given document's value with a stored reference value.
+   */
+  private interface ValueComparator {
+    int compare(int docID) throws IOException;
+  }
+
+  private static ValueComparator loadComparator(SortField sortField,
+                                                long topValue,
+                                                LeafReaderContext context) throws IOException {
+    @SuppressWarnings("unchecked")
+    FieldComparator<Long> fieldComparator = (FieldComparator<Long>) sortField.getComparator(1, 0);
+    fieldComparator.setTopValue(topValue);
+
+    LeafFieldComparator leafFieldComparator = fieldComparator.getLeafComparator(context);
+    int direction = sortField.getReverse() ? -1 : 1;
+
+    return doc -> {
+      int value = leafFieldComparator.compareTop(doc);
+      return direction * value;
+    };
+  }
+
+  /**
+   * A doc ID set iterator that wraps a delegate iterator and only returns doc IDs in
+   * the range [firstDocInclusive, lastDoc).
+   */
+  private static class BoundedDocSetIdIterator extends DocIdSetIterator {
```

Comment:
Do we need this class? Why not populate a DocSetIdIterator with just the relevant DocValues and return, much like what PointRangeQuery does?

Reply:
Because we don't need to create a bitset beforehand. Unlike points docvalues iterator access the values in the order of the internal document id so can use the original iterator there.

Reply:
Got it, so the idea is to have the same original iterator wrapped in multiple BoundedDocSetIdIterators, each with a different range.